File: 字符串/7. 重复的子字符串 - 459/Repeated Substring Pattern.py
# 23/03/30
# https://leetcode.cn/problems/repeated-substring-pattern/

class Solution:
    #Solution 1
    def repeatedSubstringPattern(self, s: str) -> bool:

        pointer = 1
        nxt = [0]
        lenOfPrefix = 0

        while pointer < len(s) :
            if s[pointer] == s[lenOfPrefix]:
                lenOfPrefix += 1
                nxt.append(lenOfPrefix)
                pointer += 1
            elif lenOfPrefix != 0:
                lenOfPrefix = nxt[lenOfPrefix - 1]
            else:
                nxt.append(0)
                pointer += 1

        return nxt[-1] != 0 and nxt[-1] % (len(s) - nxt[-1]) == 0

    # Solution 2
    def repeatedSubstringPattern1(self, s: str) -> bool:
        n = len(s)
        for i in range(1, n // 2 + 1):
            if n % i == 0:
                if all(s[j] == s[j - i] for j in range(i, n)):
                    return True
        
        return False

File: 字符串/7. 重复的子字符串 - 459/Repeated Substring Pattern.cpp
/*
    23/03/29
    https://leetcode.cn/problems/repeated-substring-pattern/
 */

#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    /*
        solution: KMP
        result: ✅
        TC: O(n)
        SC: O(1)
    */
    bool repeatedSubstringPattern(string s) {

        int next[s.size()];
        next[0] = 0;
        int pointer = 1;
        int lenOfPrefix = 0;

        while (pointer < s.size()) {
            if (s[pointer] == s[lenOfPrefix]) {
                lenOfPrefix += 1;
                next[pointer++] = lenOfPrefix;
            } else if (lenOfPrefix != 0) {
                lenOfPrefix = next[lenOfPrefix - 1];
            } else {
                next[pointer++] = 0;
            }
        }

        return lenOfPrefix != 0 && lenOfPrefix % (s.size() - lenOfPrefix) == 0 ? true : false ;
    }

    /*
        solution: 枚举
        result: ✅
        TC: O(n ^ 2)
        SC: O(1)
    */
    bool repeatedSubstringPattern1(string s) {

        int lenOfS = s.size();
        for (int i = 1; i < lenOfS; i++) {
            if (lenOfS % i == 0) {
                bool match = true;
                for (int j = i; j < lenOfS; j++) {
                    if (s[j] != s[j - i]) {
                        match = false;
                        break;
                    }
                }
                if (match) {
                    return true;
                }
            }
        }
        return false;
    }
};